common/oidc-client/oauth2-proxy/README.md

```diff
@@ -0,0 +1,231 @@
+  JWTs for authenticating requests, typically as Bearer Tokens in the
+  Authorization header. By default, the OIDC Issuer is the in-cluster Kubernetes
+  OIDC.
+
+
+## OpenShift
+
+This deployment of `oauth2-proxy` does not support OpenShift out of the box. To facilitate
+integration with OpenShift:
+* Utilize the OpenShift-specific distribution of `oauth2-proxy`, accessible at:
+    * [OpenShift OAuth Proxy on GitHub](https://github.com/openshift/oauth-proxy)
+* Enable RBAC for token reviews by incorporating the `rbac.tokenreviews.yaml` file into
+  your kustomize. For more details, refer to the relevant comment in `kustomization.yaml`.
+
+## CloudFlare
+
+CloudFlare requires that certain static, standard web browser assets are accessible without
+user authentication. This is crucial because CloudFlare aims to cache these assets for
+enhanced performance. If these assets necessitate user authentication, CloudFlare robots
+will be redirected to the authentication page, potentially causing access issues with the
+Kubeflow instance behind CloudFlare.
+
+This issue can be resolved by defining a set of assets in the Istio `AuthorizationPolicy`
+that do not require authentication. An example `AuthorizationPolicy` for this purpose is
+provided in the file `authorizationpolicy.istio-ingressgateway-oauth2-proxy-cloudflare.yaml`.
+
+## Explaining the Auth Routine
+
+1. Istio is configured with the `envoyExtAuthzHttp` extension provider pointing to the
+   `oauth2-proxy` service. This configuration enables the use of this extension in
+   `AuthorizationPolicy` for adding external authorization to the service mesh.
+2. The Istio service mesh has an `AuthorizationPolicy` named `istio-ingressgateway-oauth2-proxy`
+   in the `istio-system` (Istio root) namespace. This policy is set with `action: CUSTOM` and
+   specifies the `oauth2-proxy` provider. Consequently, every request to the
+   `istio-ingressgateway` must pass through the external authorization service, `oauth2-proxy`.
+3. `oauth2-proxy` decides based on the cookie named `oauth2_proxy_kubeflow`. If this cookie
+   is absent, expired, or invalid, `oauth2-proxy` redirects to the configured OIDC provider,
+   typically `dex`. The authentication redirect includes:
+    * A `redirect_uri` for redirecting the user post-authentication,
+    * A code challenge to guard against interception and replay attacks,
+    * A state parameter to validate the authorization response's authenticity and
+      ensure it originates from the initial request.
+
+   Post-authentication, a cookie is set in the user's browser, used by Istio and `oauth2-proxy`
+   for authorizing requests.
+
+   `oauth2-proxy` also checks for an `Authorization` header with a JWT bearer token. It trusts
+   JWTs issued by the configured OIDC provider (default `dex`) and can be configured to trust
+   additional JWTs. If a valid JWT is present, `oauth2-proxy` forwards the request to Istio
+   along with the authorization header.
```

Comment:
Where does the header come from? Is it the kubeflow-userid header? How is the kubeflow-userid header derived from the JWT in the bearer token header? This all must be in detail in the big diagram.